src/python/WMCore/MicroService/Unified/Transferor.py

```diff
@@ -72,88 +86,290 @@ def close(self):
+        records = self.store.records()
+        for rec in records:
+            if rec['tid'] == tid:
+                return rec['request']
+
+    def getTransferStatuses(self, tids):
+        "get transfer ids from backend"
+        # transfer docs on backend has the following form
+        # https://gist.github.com/amaltaro/72599f995b37a6e33566f3c749143154
+        statuses = []
+        records = self.store.records()
+        store_tids = [r.get('tid', 0) for r in records]
+        for tid in tids:
+            if tid in store_tids:
+                for rec in records:
+                    if rec.get('tid', 0) == tid:
+                       statuses.append(rec['status'])
+                       break
+            else:
+                # TODO: I need to find request name from transfer ID
+                #status = self.checkSubscription(request)
+                status = 100
+                statuses.append(status)
+        return statuses
+
+    def requestCampaign(self, req):
+        "Return request campaign"
+        pass
+
+    def requestConfiguration(self, req):
+        "Return request configuration"
+        pass
+
+    def pileupTransfersCompleted(self, tids):
+        "Check if pileup transfers are completed"
+        # TODO: add implementation
+        return False
+
+    def checkSubscription(self, request):
         "Send request to Phedex and return status of request subscription"
         sdict = {}
         rdict = self.store.get(request)
         if not rdict: # request is gone
             return 100
         for dataset in rdict.get('datasets'):
             data = self.phedex.subscriptions(dataset=dataset, group=self.group)
-            if self.verbose:
-                print("### dataset", dataset, "group", self.group)
-                print("### subscription", data)
+            self.logger.debug("### dataset %s group %s" % (dataset, self.group))
+            self.logger.debug("### subscription %s" % data)
             for row in data['phedex']['dataset']:
                 if row['name'] != dataset:
                     continue
                 nodes = [s['node'] for s in row['subscription']]
                 rNodes = rdict.get('sites')
-                if self.verbose:
-                    print("### nodes", nodes, rNodes)
+                self.logger.debug("### nodes %s %s" % (nodes, rNodes))
                 subset = set(nodes) & set(rNodes)
                 if subset == set(rNodes):
                     sdict[dataset] = 1
                 else:
                     pct = float(len(subset))/float(len(set(rNodes)))
                     sdict[dataset] = pct
-        if self.verbose:
-            print("### sdict", sdict)
+        self.logger.debug("### sdict %s" % sdict)
         tot = len(sdict.keys())
         if not tot:
             return -1
         # return percentage of completion
         return round(float(sum(sdict.values()))/float(tot), 2) * 100
 
+    def checkTransferRequests(self, request):
```

Comment:
I'm not sure I understand the difference between these 3 methods: getTransferStatuses, checkSubscription and this one. To me they seem to have the same goal, checking the status of the transfer requests from the cache or querying the data-service to get up-to-date status.

Reply:
so far it is placeholder. We may need all 3 for different purposes:
- getTransferStatuses will fetch doc from couchdb and return transfer statuses
- checkSubscription will ask upstream server
- checkTransferRequest will use both above if necessary

I don't think we should worry right now about not implemented function, instead you should evaluate place where it is used.

Reply:
I just want to avoid with code that "we may need" that is actually never needed and just adds up to code maintenance and complexity. In general, I'd say don't implement something unless you're sure it will be needed (this is something I've been fighting in WMCore over these many past years).